### PR TITLE
PWGHF: add single muon analysis task

### PR DESCRIPTION
**Pull request by MaolinZH**
This task is derived from the DQ framework and it is used to extract the observables on single muons (pT, eta, sign of the muon, DCA,  etc) needed for the HF-muon analysis. It is indeed linked to HF since the final goal is to measure muons from charm decays and muons from beauty decays, separately from the information estracted with this task.

### CODE REVIEW

**Comment by vkucera** [comment]
Hi @MaolinZH , why did you close to original PR and made a new one?

**Comment by MaolinZH** [comment]
> Hi @MaolinZH , why did you close to original PR and made a new one?

Sorry, I am not familiar with the procedure. I thought I need to make a new PR since the commit directory is changed.

**Comment by vkucera** [comment]
> > Hi @MaolinZH , why did you close to original PR and made a new one?
> 
> Sorry, I am not familiar with the procedure. I thought I need to make a new PR since the commit directory is changed.

The PR is automatically updated with whatever you push into your branch so there is no point closing it and opening a new one when making changes.
Please see the [PR review](https://aliceo2group.github.io/analysis-framework/docs/advanced-specifics/pwghf.html#pr-review) section of the instructions.

**Comment by MaolinZH** [comment]
> 



> > > Hi @MaolinZH , why did you close to original PR and made a new one?
> > 
> > 
> > Sorry, I am not familiar with the procedure. I thought I need to make a new PR since the commit directory is changed.
> 
> The PR is automatically updated with whatever you push into your branch so there is no point closing it and opening a new one when making changes. Please see the [PR review](https://aliceo2group.github.io/analysis-framework/docs/advanced-specifics/pwghf.html#pr-review) section of the instructions.

Thank you, I have read the instructions but I may have made some mistake since this is my first commit. Could you please tell me if something else is missing before to get approval?

File: PWGHF/HFL/HFmAnalysis.cxx
Comment (vkucera) [inline PWGHF/HFL/HFmAnalysis.cxx:14]: There should be a `\brief` line. See the other files.

File: PWGHF/HFL/HFmAnalysis.cxx
Comment (vkucera) [inline PWGHF/HFL/HFmAnalysis.cxx:35]: I think there are still many unnecessary includes. For example, I don't see why you need BasicCCDBManager, THashList, vector, memory, THnSparse, TDatabasePDG. Please make sure you include only what is needed.

File: PWGHF/HFL/HFmAnalysis.cxx
Comment (vkucera) [inline PWGHF/HFL/HFmAnalysis.cxx:43]: Fix the namespace name.

File: PWGHF/HFL/HFmAnalysis.cxx
Comment (vkucera) [inline PWGHF/HFL/HFmAnalysis.cxx:59]: Fix the names.

File: PWGHF/HFL/HFmAnalysis.cxx
Comment (vkucera) [inline PWGHF/HFL/HFmAnalysis.cxx:69]: Fix the names

File: PWGHF/HFL/HFmAnalysis.cxx
Comment (vkucera) [inline PWGHF/HFL/HFmAnalysis.cxx:66]:
```suggestion
  Configurable<float> fConfigSoftwareTrigger{"cfgsoftwaretrigger", VarManager::kIsMuonSingleLowPt7, "software trigger flag"};
```
Fix the typo

File: PWGHF/HFL/HFmAnalysis.cxx
Comment (vkucera) [inline PWGHF/HFL/HFmAnalysis.cxx:101]: Separate `Produces` from the function with a blank line.
Add documentation of the function.

File: PWGHF/HFL/HFmAnalysis.cxx
Comment (vkucera) [inline PWGHF/HFL/HFmAnalysis.cxx:63]: `EventSelection` is too generic. Use a more specific name that will not get easily confused with the event selection task.

File: PWGHF/HFL/HFmAnalysis.cxx
Comment (vkucera) [inline PWGHF/HFL/HFmAnalysis.cxx:149]: Replace `a` with `axis` in the names.

File: PWGHF/HFL/HFmAnalysis.cxx
Comment (vkucera) [inline PWGHF/HFL/HFmAnalysis.cxx:232]: Add documentation

File: PWGHF/HFL/HFmAnalysis.cxx
Comment (vkucera) [inline PWGHF/HFL/HFmAnalysis.cxx:245]:
```suggestion
      auto mcParticle = track.mcParticle();
```

File: PWGHF/HFL/HFmAnalysis.cxx
Comment (vkucera) [inline PWGHF/HFL/HFmAnalysis.cxx:260]:
```suggestion
      trackParCov.propagateToZlinear(event.posZ());
```

File: PWGHF/HFL/HFmAnalysis.cxx
Comment (vkucera) [inline PWGHF/HFL/HFmAnalysis.cxx:250]:
```suggestion
      std::vector<double> vecCovs{
```

File: PWGHF/HFL/HFmAnalysis.cxx
Comment (vkucera) [inline PWGHF/HFL/HFmAnalysis.cxx:326]: Add documentation

**Review by vkucera — CHANGES_REQUESTED** [review]
Many variables have bad name. Please see the [naming conventions](https://rawgit.com/AliceO2Group/CodingGuidelines/master/naming_formatting.html).

File: PWGHF/HFL/HFmAnalysis.cxx
Comment (vkucera) [inline PWGHF/HFL/HFmAnalysis.cxx:14]: The backslash denotes a Doxygen keyword. `\This` is not one.

File: PWGHF/HFL/HFmAnalysis.cxx
Comment (vkucera) [inline PWGHF/HFL/HFmAnalysis.cxx:56]: Please see the [guidelines](https://rawgit.com/AliceO2Group/CodingGuidelines/master/comments_guidelines.html?showone=Functions#Functions) to see how to document functions.

File: PWGHF/HFL/HFmAnalysis.cxx
Comment (vkucera) [inline PWGHF/HFL/HFmAnalysis.cxx:58]: This is still too ambiguous. There are other event selections in other HF code. I would suggest:
```suggestion
struct HfmEventSelection {
```

File: PWGHF/HFL/HFmAnalysis.cxx
Comment (vkucera) [inline PWGHF/HFL/HFmAnalysis.cxx:64]: `Switch` doesn't say anything about what values the parameters can have and their meaning.
Group related parameters.
```suggestion
  Configurable<bool> applySoftwareTrigger{"applySoftwareTrigger", false, "whether to apply the software trigger"};
  Configurable<float> softwareTrigger{"softwareTrigger", VarManager::kIsMuonSingleLowPt7, "software trigger flag"};
  Configurable<bool> applyCutZVtx{"applyCutZVtx", false, "whether to apply the zVtx cuts"};
  Configurable<float> zVtxMin{"zVtxMin", -10., "min. z of primary vertex [cm]"};
  Configurable<float> zVtxMax{"zVtxMax", 10., "max. z of primary vertex [cm]"};
```

**Review by vkucera — CHANGES_REQUESTED** [review]

File: PWGHF/HFL/HFmAnalysis.cxx
Comment (vkucera) [reply PWGHF/HFL/HFmAnalysis.cxx:64]: Btw, is `VarManager::kIsMuonSingleLowPt7` a `float`?

Comment (MaolinZH) [reply PWGHF/HFL/HFmAnalysis.cxx]: It should be an int, my fault.I will change it.

File: PWGHF/HFL/HFmAnalysis.cxx
Comment (vkucera) [inline PWGHF/HFL/HFmAnalysis.cxx:59]: Still the same typo.
```suggestion
  Configurable<int> softwareTrigger{"softwareTrigger", VarManager::kIsMuonSingleLowPt7, "software trigger flag"};
```

**Review by vkucera — CHANGES_REQUESTED** [review]
I see you're making changes commit by commit. That's fine. Just let me know when you're done.
- There are still variables with bad names.
- The file should be located in a proper directory, following the common directory structure.
- Make a corresponding PR in the validation framework that allows to run the task.

File: PWGHF/HFL/HFmAnalysis.cxx
Comment (MaolinZH) [reply PWGHF/HFL/HFmAnalysis.cxx:59]: Thanks for you suggestions and I do need to apology for these continuous commits. I will learn how to make it more efficient.
I am trying to find the variables with bad names and I will fix them, but there are some variables with bad names (like fHistMan, fValues) I found working well in PWGDQ codes, do I need to fix them too?
As for the directory structure. Is it okay to create a Tasks folders under the HFL folder and move the code there ? O2Physics/PWGHF/HFL/Tasks/
As I understand from the presentation from last year the validation framework is for testing and validation (w.r.t Run2).
 https://indico.cern.ch/event/1101005/contributions/4632407/attachments/2360172/4028725/2021-12-07_HF-O2-hackathon_O2.pdf
 Is it required for running this code on the AO2Ds ? Incase I misunderstood it, please let me know. 

 Thanks a lot for your help.

Comment (vkucera) [reply PWGHF/HFL/HFmAnalysis.cxx:59]: - I don't know if in PWG-DQ they have some conventions for the meaning of the `f` prefix, but to me it doesn't have any obvious purpose. It looks more like imported from an AliPhysics class but I might be wrong.
- Putting the code in the `HFL/Task` directory is fine unless it is meant to produce tables processed by another task.
- The validation framework is a convenient tool to run the HF analyses locally (as described in the PWG-HF documentation). It is recommended to add a new task there so that anyone can easily run it.

File: PWGHF/HFL/HFmAnalysis.cxx
Comment (vkucera) [reply PWGHF/HFL/HFmAnalysis.cxx]: This has not been addressed yet.

Comment (MaolinZH) [reply PWGHF/HFL/HFmAnalysis.cxx]: Sorry for my dullness but I do not get your point. Do you mean the name of defineHistograms need to be fixed or something else?

Comment (vkucera) [reply PWGHF/HFL/HFmAnalysis.cxx]: I don't mean the name. I mean that the documentation should be formatted properly with the Doxygen syntax, as explained in the link.

File: PWGHF/HFL/Tasks/HFmAnalysis.cxx
Comment (vkucera) [inline PWGHF/HFL/Tasks/HFmAnalysis.cxx:122]: Mentioning AO2D is unnecessary, since you cannot run on anything else.
```suggestion
  void processEvent(MyCollisions::iterator const& event, aod::BCs const& bcs)
  {
    runEventSel<gEventFillMap>(event, bcs);
  }

  void processEventMC(MyMcCollisions::iterator const& event, aod::BCs const& bcs)
  {
    runEventSel<gEventFillMap>(event, bcs);
  }

  PROCESS_SWITCH(HfmEventSelection, processEvent, "run event selection with real data", true);
  PROCESS_SWITCH(HfmEventSelection, processEventMC, "run event selection with MC data", false);
```

File: PWGHF/HFL/Tasks/HFmAnalysis.cxx
Comment (vkucera) [inline PWGHF/HFL/Tasks/HFmAnalysis.cxx:181]: `Data` is redundant.
```suggestion
  void runMuonSel(TEvent const& event, aod::BCs const& bcs, TMuons const& tracks)
```

File: PWGHF/HFL/Tasks/HFmAnalysis.cxx
Comment (MaolinZH) [reply PWGHF/HFL/Tasks/HFmAnalysis.cxx:313]: Thank you for the comments. The original purpose of the "AO2D" suffix is to distinguish between AO2D input and reducedAOD input which may be added in the future. I will fix them since it can not be applied to reduced AOD files at present, but I probably need to change them again afterwards.

Comment (vkucera) [reply PWGHF/HFL/Tasks/HFmAnalysis.cxx:313]: OK, I see. Thanks for the clarification. In that case, you can for example add a suffix `Reduced` to the names of the new functions when you add them.

**Review by vkucera — DISMISSED** [review]

**Comment by vkucera** [comment]
@TimoWilken Some checks are completely missing from the list. Do you know why that happened?

**Comment by vkucera** [comment]
@MaolinZH I see that the last commit changed the DCA calculation significantly. Can you please comment on that?

**Review by vkucera — APPROVED** [review]

**Pull request merged**
